### ISSUE TITLE
Service request: Signal DBus Service

### ISSUE THREAD

**Issue by matthew-piziak** [NixOS/nixpkgs#62102]
As described on the [signal-cli Wiki](https://github.com/AsamK/signal-cli/wiki/DBus-service). 

> signal-cli can run in daemon mode and provides an experimental dbus interface.

This specifically requires `jni/unix-java.so`, and requires a separate user and a configuration file.

The suggested manual process goes something like this:
```
cp data/org.asamk.Signal.conf /etc/dbus-1/system.d/
cp data/org.asamk.Signal.service /usr/share/dbus-1/system-services/
cp data/signal.service /etc/systemd/system/
sed -i -e "s|%dir%|<INSERT_INSTALL_PATH>|" -e "s|%number%|<INSERT_YOUR_NUMBER>|" /etc/systemd/system/signal.service
systemctl daemon-reload
systemctl enable signal.service
systemctl reload dbus.service
```

But it would be nice to have it packaged up as a Nix service.

**Comment by matthew-piziak** [comment]
Emacs also has a DBus interface, so with [DBus](https://www.gnu.org/software/emacs/manual/html_mono/dbus.html) and [LUI](https://www.emacswiki.org/emacs/LuiMode) it should be quite easy to create a robust and performant chat client for Signal. I don't know well enough how to package services though, so I hope that someone can help me there.

**Comment by aanderse** [comment]
@matthew-piziak Please go ahead and mention any specific issues you are having and we'll do our best to help.

**Comment by symbiont-matthew-piziak** [comment]
Thanks @aanderse! Right now I have `signal-cli` on an idle loop inside my Emacs process and connected up to LUI. It serves my purposes well enough so I probably won't jump on this soon. At some point I may want the additional robustness though, in which case I'll refer back to this issue as I try to get the service running. :)

**Comment by matthew-piziak** [comment]
In testing I got `Missing native library dependency for dbus service: no unix-java in java.library.path`.

How do I satisfy this dependency on NixOS?

**Comment by matthew-piziak** [comment]
I have `pkgs.libmatthew_java` installed, but it's not on the Java path. Here is how to do it on Fedora: https://github.com/AsamK/signal-cli/wiki/Setup-on-fedora-26-with-dbus. But I don't know how to do it in NixOS.

**Comment by matthew-piziak** [comment]
I tried `sudo ln -s /nix/store/c764v07frma64dnnwjniax7jmp0y54hb-libmatthew-java-0.8/lib/jni/libunix-java.so /usr/lib64/libunix-java.so` but it didn't help.

Is there a way I can run the `signal-cli` executable where I add the nix path to the `-Djava.library.path`?

**Comment by matthew-piziak** [comment]
I tried adding `libmatthew_java`  to `signal-cli`'s `PATH` like so:

```
installPhase = ''
      mkdir -p $out/bin
      cp -r lib $out/lib
      cp bin/signal-cli $out/bin/signal-cli
      wrapProgram $out/bin/signal-cli \
      --prefix PATH : ${lib.makeBinPath [ pkgs.jre_headless pkgs.libmatthew_java ]} \
      --set JAVA_HOME ${pkgs.jre_headless}
    '';
```

But I get the same error when trying to run `signal-cli` as daemon. What I am doing wrong?

**Comment by matthew-piziak** [comment]
I tried the following derivation, 
```
{ config, pkgs, lib, ... }:
let
  signal-cli = pkgs.signal-cli.overrideAttrs (attrs: {
    buildInputs = with pkgs; [ makeWrapper libmatthew_java ];
    installPhase = ''
      mkdir -p $out/bin
      cp -r lib $out/lib
      cp bin/signal-cli $out/bin/signal-cli
      export JAVA_OPTS='-Djava.library.path ${lib.makeBinPath [ pkgs.libmatthew_java ]}'
      wrapProgram $out/bin/signal-cli \
        --prefix PATH : ${lib.makeBinPath [ pkgs.jre_headless ]} \
        --set JAVA_HOME ${pkgs.jre_headless}
    '';
  });
in {
  environment.systemPackages = [ signal-cli ]
    ++ (with pkgs; [ qrencode shared-mime-info signal-desktop ]);
}
```
…but it fails with 
```
Error: Could not find or load main class .nix.store.h763fyxacqas0xc40b1i08dxq0dy9c07-libmatthew-java-0.8.bin
```

**Comment by matthew-piziak** [comment]
I've broken out the specific `signal-cli daemon` bug into a separate issue: https://github.com/NixOS/nixpkgs/issues/72737

**Comment by matthew-piziak** [comment]
Reproduced on Signal 0.9.2

**Comment by OliverGeneser** [comment]
@matthew-piziak Did you end up getting dbus to work?

**Comment by matthew-piziak** [comment]
@OliverGeneser Yes, `signal-cli --dbus`  for me on version 0.13.9! But I have not reattempted to configure it as a Nix service.
